# Problematic Internet Use | Part A : Data Cleaning

Problematic Internet Use by the Child Mind Institute is a Kaggle Competition for predicting the level of problematic internet usage exhibited by children and adolescents, based on their physical activity, fitness data and other demographic indicators. At the end of this project, we will hopefully have developed a predictive model to identify early signs of problematic internet use. Identifying these patterns can help trigger interventions to encourage healthier digital habits.

The link to the dataset and the competition can be found [here](https://www.kaggle.com/competitions/child-mind-institute-problematic-internet-use/overview).

## Importing Dependencies

We will import all dependencies for this part of the project in the cell below.

In [30]:
import pandas as pd
import numpy as np
import os
import matplotlib.pyplot as plt
import matplotlib.gridspec as gridspec
import seaborn as sns
import warnings

warnings.filterwarnings('ignore', category=FutureWarning)
sns.set_theme(style="whitegrid")

## Tabular Data (Instruments)

The tabular data in `train.csv` and `test.csv` comprises measurements from a variety of instruments. These instruments are:


- `Demographics` - Information about age and sex of participants.
- `Internet Use` - Number of hours of using computer/internet per day.
- `Children's Global Assessment Scale` - Numeric scale used by mental health clinicians to rate the general functioning of youths under the age of 18.
- `Physical Measures` - Collection of blood pressure, heart rate, height, weight and waist, and hip measurements.
- `FitnessGram Vitals and Treadmill` - Measurements of cardiovascular fitness assessed using the NHANES treadmill protocol.
- `FitnessGram Child` - Health related physical fitness assessment measuring five different parameters including aerobic capacity, muscular strength, muscular endurance,flexibility, and body composition.
- `Bio-electric Impedance Analysis` - Measure of key body composition elements, including BMI, fat, muscle, and water content.
- `Physical Activity Questionnaire` - Information about children's participation in vigorous activities over the last 7 days.
- `Sleep Disturbance Scale` - Scale to categorize sleep disorders in children.
- `Actigraphy` - Objective measure of ecological physical activity through a research-grade biotracker.
- `Parent-Child Internet Addiction Test` - 20-item scale that measures characteristics and behaviors associated with compulsive use of the Internet including compulsivity, escapism, and dependency.
- `Severity Impairment Index` - Based on the raw score of the Parent-Child Internet Addiction Test (0-30=None; 31-49=Mild; 50-79=Moderate; 80-100=Severe).

The fields within each instrument are described in `data_dictionary.csv`. This gives us an exact idea of what is present within the dataset:

| Instrument                         | Field                          | Description                   |
|------------------------------------|--------------------------------|-------------------------------|
| Identifier | `id` | Participant's ID |
| Demographics | `Basic_Demos-Enroll_Season` | Season of enrollment |
| Demographics | `Basic_Demos-Age` | Age of participant |
| Demographics | `Basic_Demos-Sex` | Sex of participant |
| Children's Global Assessment Scale | `CGAS-Season` | Season of participation |
| Children's Global Assessment Scale | `CGAS-CGAS_Score` | Children's Global Assessment Scale Score |
| Physical Measures | `Physical-Season` | Season of participation |
| Physical Measures | `Physical-BMI` | Body Mass Index (kg/m^2) |
| Physical Measures | `Physical-Height` | Height (in) |
| Physical Measures | `Physical-Weight` | Weight (lbs) |
| Physical Measures | `Physical-Waist_Circumference` | Waist circumference (in) |
| Physical Measures | `Physical-Diastolic_BP` | Diastolic BP (mmHg) |
| Physical Measures | `Physical-HeartRate` | Heart rate (beats/min) |
| Physical Measures | `Physical-Systolic_BP` | Systolic BP (mmHg) |
| FitnessGram Vitals and Treadmill | `Fitness_Endurance-Season` | Season of participation |
| FitnessGram Vitals and Treadmill | `Fitness_Endurance-Max_Stage` | Maximum stage reached |
| FitnessGram Vitals and Treadmill | `Fitness_Endurance-Time_Mins` | Exact time completed: Minutes |
| FitnessGram Vitals and Treadmill | `Fitness_Endurance-Time_Sec` | Exact time completed: Seconds |
| FitnessGram Child | `FGC-Season` | Season of participation |
| FitnessGram Child | `FGC-FGC_CU` | Curl up total |
| FitnessGram Child | `FGC-FGC_CU_Zone` | Curl up fitness zone |
| FitnessGram Child | `FGC-FGC_GSND` | Grip Strength total (non-dominant) |
| FitnessGram Child | `FGC-FGC_GSND_Zone` | Grip Strength fitness zone (non-dominant) |
| FitnessGram Child | `FGC-FGC_GSD` | Grip Strength total (dominant) |
| FitnessGram Child | `FGC-FGC_GSD_Zone` | Grip Strength fitness zone (dominant) |
| FitnessGram Child | `FGC-FGC_PU` | Push-up total |
| FitnessGram Child | `FGC-FGC_PU_Zone` | Push-up fitness zone |
| FitnessGram Child | `FGC-FGC_SRL` | Sit & Reach total (left side) |
| FitnessGram Child | `FGC-FGC_SRL_Zone` | Sit & Reach fitness zone (left side) |
| FitnessGram Child | `FGC-FGC_SRR` | Sit & Reach total (right side) |
| FitnessGram Child | `FGC-FGC_SRR_Zone` | Sit & Reach fitness zone (right side) |
| FitnessGram Child | `FGC-FGC_TL` | Trunk lift total |
| FitnessGram Child | `FGC-FGC_TL_Zone` | Trunk lift fitness zone |
| Bio-electric Impedance Analysis | `BIA-Season` | Season of participation |
| Bio-electric Impedance Analysis | `BIA-BIA_Activity_Level_num` | Activity Level |
| Bio-electric Impedance Analysis | `BIA-BIA_BMC` | Bone Mineral Content |
| Bio-electric Impedance Analysis | `BIA-BIA_BMI` | Body Mass Index |
| Bio-electric Impedance Analysis | `BIA-BIA_BMR` | Basal Metabolic Rate |
| Bio-electric Impedance Analysis | `BIA-BIA_DEE` | Daily Energy Expenditure |
| Bio-electric Impedance Analysis | `BIA-BIA_ECW` | Extracellular Water |
| Bio-electric Impedance Analysis | `BIA-BIA_FFM` | Fat Free Mass |
| Bio-electric Impedance Analysis | `BIA-BIA_FFMI` | Fat Free Mass Index |
| Bio-electric Impedance Analysis | `BIA-BIA_FMI` | Fat Mass Index |
| Bio-electric Impedance Analysis | `BIA-BIA_Fat` | Body Fat Percentage |
| Bio-electric Impedance Analysis | `BIA-BIA_Frame_num` | Body Frame |
| Bio-electric Impedance Analysis | `BIA-BIA_ICW` | Intracellular Water |
| Bio-electric Impedance Analysis | `BIA-BIA_LDM` | Lean Dry Mass |
| Bio-electric Impedance Analysis | `BIA-BIA_LST` | Lean Soft Tissue |
| Bio-electric Impedance Analysis | `BIA-BIA_SMM` | Skeletal Muscle Mass |
| Bio-electric Impedance Analysis | `BIA-BIA_TBW` | Total Body Water |
| Physical Activity Questionnaire (Adolescents) | `PAQ_A-Season` | Season of participation |
| Physical Activity Questionnaire (Adolescents) | `PAQ_A-PAQ_A_Total` | Activity Summary Score (Adolescents) |
| Physical Activity Questionnaire (Children) | `PAQ_C-Season` | Season of participation |
| Physical Activity Questionnaire (Children) | `PAQ_C-PAQ_C_Total` | Activity Summary Score (Children) |
| Parent-Child Internet Addiction Test | `PCIAT-Season` | Season of participation |
| Parent-Child Internet Addiction Test | `PCIAT-PCIAT_01` | How often does your child disobey time limits you set for online use? |
| Parent-Child Internet Addiction Test | `PCIAT-PCIAT_02` | How often does your child neglect household chores to spend more time online? |
| Parent-Child Internet Addiction Test | `PCIAT-PCIAT_03` | How often does your child prefer to spend time online rather than with the rest of your family? |
| Parent-Child Internet Addiction Test | `PCIAT-PCIAT_04` | How often does your child form new relationships with fellow online users? |
| Parent-Child Internet Addiction Test | `PCIAT-PCIAT_05` | How often do you complain about the amount of time your child spends online? |
| Parent-Child Internet Addiction Test | `PCIAT-PCIAT_06` | How often do your child's grades suffer because of the amount of time he or she spends online? |
| Parent-Child Internet Addiction Test | `PCIAT-PCIAT_07` | How often does your child check his or her e-mail before doing something else? |
| Parent-Child Internet Addiction Test | `PCIAT-PCIAT_08` | How often does your child seem withdrawn from others since discovering the Internet? |
| Parent-Child Internet Addiction Test | `PCIAT-PCIAT_09` | How often does your child become defensive or secretive when asked what he or she does online? |
| Parent-Child Internet Addiction Test | `PCIAT-PCIAT_10` | How often have you caught your child sneaking online against your wishes? |
| Parent-Child Internet Addiction Test | `PCIAT-PCIAT_11` | How often does your child spend time alone in his or her room playing on the computer? |
| Parent-Child Internet Addiction Test | `PCIAT-PCIAT_12` | How often does your child receive strange phone calls from new "online" friends? |
| Parent-Child Internet Addiction Test | `PCIAT-PCIAT_13` | How often does your child snap, yell, or act annoyed if bothered while online? |
| Parent-Child Internet Addiction Test | `PCIAT-PCIAT_14` | How often does your child seem more tired and fatigued than he or she did before the Internet came along? |
| Parent-Child Internet Addiction Test | `PCIAT-PCIAT_15` | How often does your child seem preoccupied with being back online when off-line? |
| Parent-Child Internet Addiction Test | `PCIAT-PCIAT_16` | How often does your child throw tantrums with your interference about how long he or she spends online? |
| Parent-Child Internet Addiction Test | `PCIAT-PCIAT_17` | How often does your child choose to spend time online rather than doing once enjoyed hobbies and/or outside interests? |
| Parent-Child Internet Addiction Test | `PCIAT-PCIAT_18` | How often does your child become angry or belligerent when your place time limits on how much time he or shes is allowed to spend online? |
| Parent-Child Internet Addiction Test | `PCIAT-PCIAT_19` | How often does your child choose to spend more time online than going out with friends? |
| Parent-Child Internet Addiction Test | `PCIAT-PCIAT_20` | How often does your child feel depressed, moody, or nervous when off-line which seems to go away once back online? |
| Parent-Child Internet Addiction Test | `PCIAT-PCIAT_Total` | Total Score |
| Sleep Disturbance Scale | `SDS-Season` | Season of participation |
| Sleep Disturbance Scale | `SDS-SDS_Total_Raw` | Total Raw Score |
| Sleep Disturbance Scale | `SDS-SDS_Total_T` | Total T-Score |
| Internet Use | `PreInt_EduHx-Season` | Season of participation |
| Internet Use | `PreInt_EduHx-computerinternet_hoursday` | Hours of using computer/internet |


The aim of this competition is to predict the Severity Impairment Index (`sii`), which measures the level of problematic internet use among children and adolescents, based on physical activity data and other features.

As previously mentioned, `sii` is derived from `PCIAT-PCIAT_Total`, the sum of scores from the Parent-Child Internet Addiction Test (PCIAT: 20 questions, scored 0-5).

Target Variable (`sii`) is defined as:
- 0: None (`PCIAT-PCIAT_Total` from 0 to 30)
- 1: Mild (`PCIAT-PCIAT_Total` from 31 to 49)
- 2: Moderate (`PCIAT-PCIAT_Total` from 50 to 79)
- 3: Severe (`PCIAT-PCIAT_Total` 80 and more)

This makes `sii` an ordinal categorical variable with four levels, where the order of categories is meaningful.

Types of Machine Learning Problem we can use with `sii` as a target:

- Ordinal classification (ordinal logistic regression, models with custom ordinal loss functions)
- Multiclass classification (treat sii as a nominal categorical variable without considering the order)
- Regression (ignore the discrete nature of categories and treat sii as a continuous variable, then round prediction)
- Custom (e.g. loss functions that penalize errors based on the distance between categories)

We can also use `PCIAT-PCIAT_Total` as a continuous target variable, and implement regression on `PCIAT-PCIAT_Total` and then map predictions to `sii` categories.

But let's not get ahead of ourselves. Firstly, we should do some data cleaning while obviously taking great care to not to incur any imputations, by keeping data ethics in mind at every step.


## Data Preview

Loading in the data:

In [ ]:
# Load data
train = pd.read_csv('train.csv')
data_dict = pd.read_csv('data_dictionary.csv')

Briefly displaying what we are working with:

In [ ]:
# Display the tabular data
display(train.head())
print("Train shape:", train.shape)

# Store the initial number of rows in the training data
num_rows_start = train.shape[0]

id Basic_Demos-Enroll_Season  Basic_Demos-Age  Basic_Demos-Sex  \
0  00008ff9                      Fall                5                0   
1  000fd460                    Summer                9                0   
2  00105258                    Summer               10                1   
3  00115b9f                    Winter                9                0   
4  0016bb22                    Spring               18                1   

  CGAS-Season  CGAS-CGAS_Score Physical-Season  Physical-BMI  Physical-Height  \
0      Winter             51.0            Fall     16.877316             46.0   
1         NaN              NaN            Fall     14.035590             48.0   
2        Fall             71.0            Fall     16.648696             56.5   
3        Fall             71.0          Summer     18.292347             56.0   
4      Summer              NaN             NaN           NaN              NaN   

   Physical-Weight  ...  PCIAT-PCIAT_18  PCIAT-PCIAT_19  PCIAT-PCIAT_20  \
0             50.8  ...             4.0             2.0             4.0   
1             46.0  ...             0.0             0.0             0.0   
2             75.6  ...             2.0             1.0             1.0   
3             81.6  ...             3.0             4.0             1.0   
4              NaN  ...             NaN             NaN             NaN   

   PCIAT-PCIAT_Total SDS-Season  SDS-SDS_Total_Raw  SDS-SDS_Total_T  \
0               55.0        NaN                NaN              NaN   
1                0.0       Fall               46.0             64.0   
2               28.0       Fall               38.0             54.0   
3               44.0     Summer               31.0             45.0   
4                NaN        NaN                NaN              NaN   

   PreInt_EduHx-Season PreInt_EduHx-computerinternet_hoursday  sii  
0                 Fall                                    3.0  2.0  
1               Summer                                    0.0  0.0  
2               Summer                                    2.0  0.0  
3               Winter                                    0.0  1.0  
4                  NaN                                    NaN  NaN  

[5 rows x 82 columns]

Train shape: (3960, 82)


## Verification of Target Variables

As shown previously, the Parent-Child Internet Addiction Test contains 20 questions (`PCIAT-PCIAT_01` to `PCIAT-PCIAT_20`), each assessing a different aspect of a child's behavior related to internet use. The questions are answered on a scale (from 0 to 5), and the total score provides an indication of the severity of internet addiction.

We also have season of participation in `PCIAT-Season` and total Score in `PCIAT-PCIAT_Total`, so there are 22 PCIAT test-related columns in total.

Let's verify that all the `PCIAT-PCIAT_Total`s align with their corresponding `sii` categories by calculating the minimum and maximum scores for each `sii` category:

In [21]:
# Finding the minimum and maximum PCIAT total score for each sii
pciat_min_max = train.groupby('sii')['PCIAT-PCIAT_Total'].agg(['min', 'max'])

# Renaming the columns
pciat_min_max = pciat_min_max.rename(
    columns={'min': 'Minimum PCIAT total Score', 'max': 'Maximum total PCIAT Score'}
)

display(pciat_min_max)
# Should be: (PCIAT Range, sii) = (0-30, 0), (31-49, 1), (50-79, 2), (80-100, 3)

Minimum PCIAT total Score  Maximum total PCIAT Score
sii                                                      
0.0                        0.0                       30.0
1.0                       31.0                       49.0
2.0                       50.0                       79.0
3.0                       80.0                       93.0

It's been stated that some of the Parent-Child Internet Addiction Test questions can be ignored by a respondent, hence there are plently of missing values in the `PCIAT-PCIAT_01` to `PCIAT-PCIAT_20` columns, but the `sii` score is still derived from the the sum of the non-NA values, leading to potentially invalid `sii` values.

In [22]:
# Filtering for just the PCIAT questions, total score and sii columns
target_variable_columns = train.filter(regex='PCIAT|sii').columns

# Gettting a new temporary dataframe with only the target variable columns
train_with_sii = train[train['sii'].notna()][target_variable_columns]

# Highlighting missing values
train_with_sii[train_with_sii.isna().any(axis=1)].head().style.applymap(
    lambda x: 'background-color: #FC0303' if pd.isna(x) else ''
)

In the 1st and 3rd rows we can see that the score for one answer is missing. And since each question is scored from 1 to 5, the total score could be up to 5 points higher and correspond to the next `sii` category (`sii` can be 0 or 1 for the first row and 1 or 2 for the third). It also seems like `PCIAT-PCIAT_Total` and `sii` get a default value of 0 when no test questions have even been answered at all, such as in the second row.

Let's check if `PCIAT-PCIAT_Total` was indeed calculated as a sum of only non-NA values in `PCIAT-PCIAT_01` to `PCIAT-PCIAT_20` columns:

In [23]:
# Gettting just the 20 PCIAT questions columns
PCIAT_cols = [f'PCIAT-PCIAT_{i+1:02d}' for i in range(20)]

# Manually recalculating the total PCIAT score without accounting for the missing values
recalc_total_score = train_with_sii[PCIAT_cols].sum(
    axis=1, skipna=True
)

# Checking if the recalculated total score matches the total score already in the dataset
(recalc_total_score == train_with_sii['PCIAT-PCIAT_Total']).all()

True

We can conclude that the `sii` score is sometimes incorrect. To fix this, we can estimate the values of missing PCIAT questions by using the median of the non-missing values (median imputation). For rows where no PCIAT questions have been answered at all, we can drop them entirely:

In [24]:
# Drop rows where all 20 PCIAT-PCIAT_* columns are nan
pciat_questions_cols = [f'PCIAT-PCIAT_{i:02d}' for i in range(1, 21)]
train = train.dropna(subset=pciat_questions_cols, how='all')

# Fill NaNs with the row-wise median for each row in the 20 PCIAT columns
for index, row in train.iterrows():
    row_median = row[pciat_questions_cols].median(skipna=True)
    train.loc[index, pciat_questions_cols] = row[pciat_questions_cols].fillna(row_median)

# Recompute PCIAT-PCIAT_Total as the sum of PCIAT-PCIAT_01 to PCIAT-PCIAT_20
train['PCIAT-PCIAT_Total'] = train[pciat_questions_cols].sum(axis=1)

# Recompute sii based on the updated PCIAT-PCIAT_Total values
def compute_sii(total):
    if 0 <= total <= 30:
        return 0  # None
    elif 31 <= total <= 49:
        return 1  # Mild
    elif 50 <= total <= 79:
        return 2  # Moderate
    else:
        return 3  # Severe
    
train['sii'] = train['PCIAT-PCIAT_Total'].apply(compute_sii)

# Displaying the first 5 rows of the target variable columns after these changes
train[target_variable_columns].head()

PCIAT-Season  PCIAT-PCIAT_01  PCIAT-PCIAT_02  PCIAT-PCIAT_03  \
0         Fall             5.0             4.0             4.0   
1         Fall             0.0             0.0             0.0   
2         Fall             5.0             2.0             2.0   
3       Summer             4.0             2.0             4.0   
5       Summer             3.0             3.0             3.0   

   PCIAT-PCIAT_04  PCIAT-PCIAT_05  PCIAT-PCIAT_06  PCIAT-PCIAT_07  \
0             0.0             4.0             0.0             0.0   
1             0.0             0.0             0.0             0.0   
2             1.0             2.0             1.0             1.0   
3             0.0             5.0             1.0             0.0   
5             0.0             2.0             1.0             0.0   

   PCIAT-PCIAT_08  PCIAT-PCIAT_09  ...  PCIAT-PCIAT_13  PCIAT-PCIAT_14  \
0             4.0             0.0  ...             4.0             4.0   
1             0.0             0.0  ...             0.0             0.0   
2             2.0             1.0  ...             1.0             1.0   
3             3.0             2.0  ...             3.0             0.0   
5             2.0             2.0  ...             3.0             3.0   

   PCIAT-PCIAT_15  PCIAT-PCIAT_16  PCIAT-PCIAT_17  PCIAT-PCIAT_18  \
0             4.0             4.0             4.0             4.0   
1             0.0             0.0             0.0             0.0   
2             1.0             0.0             2.0             2.0   
3             0.0             3.0             4.0             3.0   
5             2.0             1.0             3.0             1.0   

   PCIAT-PCIAT_19  PCIAT-PCIAT_20  PCIAT-PCIAT_Total  sii  
0             2.0             4.0               55.0    2  
1             0.0             0.0                0.0    0  
2             1.0             1.0               28.0    0  
3             4.0             1.0               44.0    1  
5             2.0             1.0               34.0    1  

[5 rows x 23 columns]

## Identifying other Missing Values

Now that our target variables are sorted, we can move onto investigating all the other features of our tabular data.

Firstly, we should assess the extent of missing data across all features. This will provide insight into which columns require imputation and which have too many missing values to be useful. Let's calculated the number and percentage of missing values for each feature:

In [26]:
# Check for missing values in each column
missing_data = train.isnull().sum().sort_values(ascending=False)
percent_missing = (missing_data / len(train)) * 100
missing_summary = pd.DataFrame({'Missing Values': missing_data, 'Percent Missing': percent_missing})
pd.set_option('display.max_rows', None)
print(missing_summary)

                                        Missing Values  Percent Missing
PAQ_A-PAQ_A_Total                                 2372        86.727605
PAQ_A-Season                                      2372        86.727605
Physical-Waist_Circumference                      2252        82.340037
Fitness_Endurance-Time_Sec                        2007        73.382084
Fitness_Endurance-Time_Mins                       2007        73.382084
Fitness_Endurance-Max_Stage                       2004        73.272395
FGC-FGC_GSND_Zone                                 1871        68.409506
FGC-FGC_GSD_Zone                                  1871        68.409506
FGC-FGC_GSD                                       1864        68.153565
FGC-FGC_GSND                                      1863        68.117002
Fitness_Endurance-Season                          1475        53.930530
PAQ_C-Season                                      1295        47.349177
PAQ_C-PAQ_C_Total                                 1295        47

For features where majority of the data is unavailable, imputing values is most likely to not add meaningful information. In fact, imputation in this case will largely provide us with a redundant or misleading feature.

Therefore, for features where more than 70% of the data is missing, such as `PAQ_A-PAQ_A_Total` and `Fitness_Endurance-Time_Sec`, we'll opt to drop these columns. 

In [27]:
# Drop columns with more than 70% missing values
train = train.drop(columns=[
    'PAQ_A-PAQ_A_Total', 
    'PAQ_A-Season',
    'Physical-Waist_Circumference',
    'Fitness_Endurance-Time_Sec', 
    'Fitness_Endurance-Time_Mins', 
    'Fitness-Endurance-Max_Stage'
], errors='ignore')

For columns with moderate amounts of missing data, we can use two different imputation methods:

- Numerical Features: For numerical columns, we can use the median to fill in missing values. Median imputation is a method used to replace missing values in numerical data with the median value of the column. This technique is especially useful when the data contains outliers, as the median is less affected by extreme values compared to the mean. 
    
    (Note that this is different from the median imputation we did for the PCIAT questions. There we used each child's answered questions to estimate the values of the non-answered questions. Since the remaining columns with missing values are largely independant of each other now, for each child with a missing value we are estimating it using the values for the other children).

- Categorical Features: For categorical columns, we can apply mode imputation. Mode imputation is a technique used to fill in missing values in categorical data by replacing them with the most frequently occurring value in the dataset (i.e., the mode). This method is particularly useful for handling missing values in features that represent categories, as it helps maintain the overall distribution of the data.


In [28]:
# List of numerical columns that need median imputation
numerical_columns = [
    'FGC-FGC_GSND_Zone', 'FGC-FGC_GSD_Zone', 
    'FGC-FGC_GSD', 'FGC-FGC_GSND', 'PAQ_C-PAQ_C_Total', 
    'BIA-BIA_Frame_num', 'BIA-BIA_Fat', 'BIA-BIA_FFMI', 'BIA-BIA_FFM', 
    'BIA-BIA_ECW', 'BIA-BIA_DEE', 'BIA-BIA_BMR', 'BIA-BIA_Activity_Level_num', 
    'BIA-BIA_FMI', 'BIA-BIA_LDM', 'BIA-BIA_ICW', 'BIA-BIA_BMI', 'BIA-BIA_LST', 
    'BIA-BIA_SMM', 'BIA-BIA_TBW', 'BIA-BIA_BMC', 
    'CGAS-CGAS_Score', 'SDS-SDS_Total_T', 'SDS-SDS_Total_Raw', 
    'Physical-Systolic_BP', 'Physical-Diastolic_BP', 'Physical-HeartRate', 
    'Physical-BMI', 'Physical-Height', 'Physical-Weight', 
    'PreInt_EduHx-computerinternet_hoursday'
]

# Apply median imputation for numerical columns
for col in numerical_columns:
    if col in train.columns and pd.api.types.is_numeric_dtype(train[col]):
        train[col] = train[col].fillna(train[col].median())

# Categorical columns that need mode imputation
categorical_columns = [
    'BIA-Season', 'CGAS-Season', 'SDS-Season', 'PCIAT-Season', 
    'Physical-Season', 'FGC-Season', 'PreInt_EduHx-Season', 
    'Fitness_Endurance-Season','Fitness_Endurance-Max_Stage','PAQ_C-Season',  
    'FGC-FGC_SRL_Zone', 'FGC-FGC_SRR_Zone', 'FGC-FGC_PU_Zone', 
    'FGC-FGC_CU_Zone', 'FGC-FGC_TL_Zone', 'FGC-FGC_SRL', 
    'FGC-FGC_SRR', 'FGC-FGC_PU', 'FGC-FGC_CU', 'FGC-FGC_TL'
]

# Apply mode imputation for categorical columns
for col in categorical_columns:
    if col in train.columns:
        train[col] = train[col].fillna(train[col].mode()[0])

After dropping the columns with excessive missing values and imputing missing data in the remaining columns, let's recheck the dataset to ensure no missing values remain. This will confirm if the dataset is fully cleaned and ready for EDA and other future tasks.

In [29]:
missing_data_post_clean = train.isnull().sum().sort_values(ascending=False)
print(missing_data_post_clean)

id                                        0
PCIAT-Season                              0
PCIAT-PCIAT_07                            0
PCIAT-PCIAT_06                            0
PCIAT-PCIAT_05                            0
PCIAT-PCIAT_04                            0
PCIAT-PCIAT_03                            0
PCIAT-PCIAT_02                            0
PCIAT-PCIAT_01                            0
PAQ_C-PAQ_C_Total                         0
BIA-BIA_FMI                               0
PAQ_C-Season                              0
BIA-BIA_TBW                               0
BIA-BIA_SMM                               0
BIA-BIA_LST                               0
BIA-BIA_LDM                               0
BIA-BIA_ICW                               0
BIA-BIA_Frame_num                         0
PCIAT-PCIAT_08                            0
PCIAT-PCIAT_09                            0
PCIAT-PCIAT_10                            0
PCIAT-PCIAT_11                            0
PreInt_EduHx-computerinternet_ho

As we can see, all missing values have been handled, with no gaps remaining in the data. Let's drop all duplicate rows too just in case.

In [ ]:
# Drop duplicate rows (if any)
train.drop_duplicates(inplace=True)

Let's see how many rows we are left with at the end:

In [41]:
# Display the number of rows before and after cleaning
print("Initial number of rows in the training data:", num_rows_start)
num_rows_end = train.shape[0]
print("Number of rows after cleaning:", num_rows_end)
print("")

# Calculate the percentage of rows removed
percent_removed = ((num_rows_start - num_rows_end) / num_rows_start) * 100
print(f"Percentage of rows removed: {percent_removed:.2f}%")

Initial number of rows in the training data: 3960
Number of rows after cleaning: 2735

Percentage of rows removed: 30.93%


Even though we lost almost a third of our dataset, we still have data for 2735 participants, which is most likely plently for drawing meaningful insights through EDA, causal inference and for training a good linear regression model. Let's go ahead and save our cleaned data as a `.csv`.

In [42]:
train.to_csv('cleaned_data.csv', index=False)

## Actigraphy Data

Other than our main, tabular dataset, we are also provided with time-series data. Certain participants agreed to wearing wrist-worn accelerometers for up to 30 days, thus providing real-time insights into participants' physical activity over an extended period. The features for each time-series are as follows:


- `id` - The patient identifier corresponding to the id field in train/test.csv.
- `step` - An integer timestep for each observation within a series.
- `X, Y, Z` - Measure of acceleration, in g, experienced by the wrist-worn watch along each standard axis.
- `enmo` - As calculated and described by the wristpy package, ENMO is the Euclidean Norm Minus One of all accelerometer signals (along each of the x-, y-, and z-axis, measured in g-force) with negative values rounded to zero. Zero values are indicative of periods of no motion. While no standard measure of acceleration exists in this space, this is one of the several commonly computed features.
- `anglez` - As calculated and described by the wristpy package, Angle-Z is a metric derived from individual accelerometer components and refers to the angle of the arm relative to the horizontal plane.
- `non-wear_flag` - A flag (0: watch is being worn, 1: the watch is not worn) to help determine periods when the watch has been removed, based on the GGIR definition, which uses the standard deviation and range of the accelerometer data.
- `light` - Measure of ambient light in lux. See ​​here for details.
- `battery_voltage` - A measure of the battery voltage in mV.
- `time_of_day` - Time of day representing the start of a 5s window that the data has been sampled over, with format `%H:%M:%S.%9f`.
- `weekday` - The day of the week, coded as an integer with 1 being Monday and 7 being Sunday.
- `quarter` - The quarter of the year, an integer from 1 to 4.
- `relative_date_PCIAT` - The number of days (integer) since the PCIAT test was administered (negative days indicate that the actigraphy data has been collected before the test was administered).

The actigraphy data in conjuction with the tabular data, helps us capture both behavioral patterns (like internet usage and sleep quality) and physical activity (tracked through both surveys and the wrist-worn accelerometers). This will prove useful during causal inference.

The time-series data is stored within the `series_train.parquet`, which contains one folder each for each participant who wore an accelerometer (indicated by their `id`), which in turn contains the data file in `.parquet` format. Let's take a brief look at the time-series data for the participant with `id:00f332d1`:

In [38]:
df = pd.read_parquet('series_train.parquet/id=00f332d1/part-0.parquet')
display(df.head())
print("Shape of the dataframe:", df.shape)

step         X         Y         Z      enmo     anglez  non-wear_flag  \
0     0 -0.051997  0.011325 -0.991420  0.026643 -86.152954            0.0   
1     1 -0.008801 -0.036264 -1.000591  0.001868 -87.851646            0.0   
2     2 -0.008904 -0.035101 -1.000357  0.001776 -87.853790            0.0   
3     3 -0.009110 -0.035204 -1.000435  0.001675 -87.850815            0.0   
4     4 -0.009020 -0.035204 -1.000709  0.001846 -87.850815            0.0   

       light  battery_voltage     time_of_day  weekday  quarter  \
0  30.000000           4185.0  57240000000000        1        2   
1  30.000000           4185.0  57245000000000        1        2   
2  30.333334           4185.0  57250000000000        1        2   
3  30.666666           4185.0  57255000000000        1        2   
4  31.000000           4185.0  57260000000000        1        2   

   relative_date_PCIAT  
0                 14.0  
1                 14.0  
2                 14.0  
3                 14.0  
4                 14.0

Shape of the dataframe: (414384, 13)


Since only folders for those participants are available who wore an accelometer, and since the data was automatically logged at every time step, we expect every `.parquet` file to be complete and not contain any missing values. However, let's still verify this:

In [39]:
# Get the filepaths of every .parquet file
parquet_paths = []
for root, dirs, files in os.walk('series_train.parquet'):
    for file in files:
        parquet_paths.append(os.path.join(root, file))

# Display the amount of parquet files
print("Number of parquet files:", len(parquet_paths))

# Open each parquet file and check for NaN values, if any are found, break the loop
nan_found = False
for file in parquet_paths:
    df = pd.read_parquet(file)
    has_nan = df.isnull().values.any()
    if has_nan:
        nan_found = True
        break

if nan_found:
    print("Nan values found in the parquet files")
else:
    print("No Nan values found in the parquet files")
    

Number of parquet files: 996
No Nan values found in the parquet files


As we have seen, the time-series data is already fully logged and complete, therefore no data cleaning or imputations are required here.

__This marks the end of Part A: Data Cleaning.__